### PR TITLE
8269685: Optimize HeapHprofBinWriter implementation

### PR DESCRIPTION
**Pull request by linzang**
This PR rewrite the implementation of the HeapHprofBinWriter, which could simplify the logic of current implementation.
please see detail description at https://bugs.openjdk.java.net/browse/JDK-8269685.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issues
 * [JDK-8269685](https://bugs.openjdk.java.net/browse/JDK-8269685): Optimize HeapHprofBinWriter implementation
 * [JDK-8262386](https://bugs.openjdk.java.net/browse/JDK-8262386): resourcehogs/serviceability/sa/TestHeapDumpForLargeArray.java timed out ⚠️ Issue is not open.

### Reviewers
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to de20431aed522661db64c56af0c30876c88507fa
 * [Alex Menkov](https://openjdk.java.net/census#amenkov) (@alexmenkov - **Reviewer**) ⚠️ Review applies to 1da3ab12569482629c115fd197ea26480ebc2c53

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4666/head:pull/4666` \
`$ git checkout pull/4666`

Update a local copy of the PR: \
`$ git checkout pull/4666` \
`$ git pull https://git.openjdk.java.net/jdk pull/4666/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4666`

View PR using the GUI difftool: \
`$ git pr show -t 4666`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4666.diff">https://git.openjdk.java.net/jdk/pull/4666.diff</a>

</details>

### CODE REVIEW

**Comment by linzang** [comment]
This PR also fix the issue https://bugs.openjdk.java.net/browse/JDK-8269685.

**Comment by linzang** [comment]
Dear All,
May I ask your help to review this PR? it fix the issue JDK-8262386.
I believe this could be a replacement for #2803.  I am planing to close that if this one looks good.

BRs,
Lin

**Comment by linzang** [comment]
Added the tests I have conducted:
    Passed test tier1 (tier2 and tier3 not executed)
    Tested to dump netbeans with heap size at ~500MB, ~700MB and 1GB, and tested w/o -gz option from 1-9. all passed.
    A workload that generate different object to fill heap, I have tested with heap usage from 1GB to 4GB with different gz options. All passed.

**Comment by sspitsyn** [comment]
Hi Lin,

You did not answer my question about empty methods in the AbstractHeapGraphWriter.java.
It seems the following methods have to be abstract instead of being empty:
```
+    protected int calculateGlobalJNIHandlesDumpRecordSize() {
+        return 0;
+    }
+
+    protected int calculateJavaThreadsDumpRecordSize() {
+        return 0;
+    }
+
+    protected int calculateOopDumpRecordSize(Oop oop) throws IOException {
+        return 0;
+    }
. . . 
    // object field writers
protected void writeReferenceField(Oop oop, OopField field)
     throws IOException {
}
protected void writeByteField(Oop oop, ByteField field)
    throws IOException {
}
protected void writeCharField(Oop oop, CharField field)
    throws IOException {
}
protected void writeBooleanField(Oop oop, BooleanField field)
    throws IOException {
}
protected void writeShortField(Oop oop, ShortField field)
    throws IOException {
}
protected void writeIntField(Oop oop, IntField field)
    throws IOException {
}
protected void writeLongField(Oop oop, LongField field)
    throws IOException {
}
protected void writeFloatField(Oop oop, FloatField field)
    throws IOException {
}
protected void writeDoubleField(Oop oop, DoubleField field)
    throws IOException {
}
protected void writeObjectFooter(Oop oop) throws IOException {
}
protected void writeHeapFooter() throws IOException {
}
protected void writeHeapRecordPrologue() throws IOException {
} 
+    protected void writeHeapRecordPrologue(int size) throws IOException {
+    }
+
protected void writeHeapRecordEpilogue() throws IOException {
}
```

**Comment by linzang** [comment]
Hi Serguei,
   So sorry that I missed your comments, it is strange that your review didn't show on github in this PR.
   I make these methods empty in AbstractHeapGraphWriter rather than abstract so that there is no need to add these empty code in HeapGXLWriter, which is a subclass of AbstractHeapGraphWriter.  I think this may help make the code change clean. 
   What do you think?

BRs,
Lin

**Comment by sspitsyn** [comment]
Lin, in order to see my inlined comment you need to look at "Files changed".
I still miss something. If these methods have to be empty then what is the purpose to have them?
Also, a couple of new empty methods are not used, so the question is why do you add them:
  calculateGlobalJNIHandlesDumpRecordSize, calculateJavaThreadsDumpRecordSize

**Comment by linzang** [comment]
Dear Serguei,

> Lin, in order to see my inlined comment you need to look at "Files changed".
I also can not find your comments here, would you like to help check whether you sumbited the inlined commit after review?

> I still miss something. If these methods have to be empty then what is the purpose to have them?

The main reason is that both `HeapHprofBinWriter` and `HeapGXLWriter` are extended from `AbstractHeapGraphWriter`, and they both invoke method `write()` in `AbstractHeapGraphWriter` for heap iteration which invoke `doObj()`.  And in method `doObj()`, it calls `writeHeapRecordPrologue(calculateOopDumpRecordSize())` which should do nothing for `HeapGXLWriter` and do some work for `HeapHprofBinWriter`.  So I made empty method `writeHeapRecordPrologue(int)` and `calculateOopDumpRecordSize()` in base class and have them implemented in `HeapHprofBinWriter` only. 

> Also, a couple of new empty methods are not used, so the question is why do you add them:
> calculateGlobalJNIHandlesDumpRecordSize, calculateJavaThreadsDumpRecordSize

Thanks for point them out, I will remove these not used methods.

BRs,
Lin

**Comment by sspitsyn** [comment]
Lin, you are right. It occurred I did not submit my comment (while I was sure that I did it).

**Comment by linzang** [comment]
Hi Serguei, 
    I update the PR to make the `calculateGlobalJNIHandlesDumpRecordSize()` method abstract. 
    For other empty ones that already there in original implementation, I will leave them as they are. do you think it is OK? 

BRs,
Lin

**Comment by sspitsyn** [comment]
> Hi Serguei,
> I update the PR to make the `calculateGlobalJNIHandlesDumpRecordSize()` method abstract.
> For other empty ones that already there in original implementation, I will leave them as they are. do you think it is OK?
> 
> BRs,
> Lin

Hi Lin,
Yes, I'm okay with that.
Thanks,
Serguei

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java [hunk 1]
```diff
@@ -419,9 +419,7 @@ public synchronized void write(String fileName) throws IOException {
                     }
                 };
             }
-            hprofBufferedOut = new SegmentedOutputStream(dataOut);
-        } else {
-            hprofBufferedOut = new SegmentedOutputStream(fos, false /* allowSegmented */);
+            hprofBufferedOut = dataOut;
```
Comment (sspitsyn) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java:422]: I do not understand why the local variable `dataOut` is needed.
The code will be simpler without it:
```
    OutputStream dataOut = fos;
     hprofBufferedOut = dataOut;
     if (useSegmentedHeapDump) {
         if (isCompression()) {
             dataOut = new GZIPOutputStream(fos) {
                 {
                     this.def.setLevel(gzLevel);
                 }
             };
         }
             hprofBufferedOut = dataOut;
     }
```
=====>
```
     hprofBufferedOut = fos;
     if (useSegmentedHeapDump) {
         if (isCompression()) {
             hprofBufferedOut = new GZIPOutputStream(fos) {
                 {
                     this.def.setLevel(gzLevel);
                 }
             };
         }
     }
```

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java [hunk 2]
```diff
@@ -472,47 +470,213 @@ public synchronized void write(String fileName) throws IOException {
...
+                klass = klass.getSuper();
+                while (klass != null) {
+                    name = klass.getName();
+                    if (name.equals(javaLangThread)) {
+                        return calculateThreadDumpRecordSize(instance);
+                    }
+                    klass = klass.getSuper();
+                }
+                return calculateInstanceDumpRecordSize(instance);
+            }
```
Comment (sspitsyn) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java:523]: The code above is really confusing.
You have these methods defined (Q: why are these methods not defined close to each other?):
```
+    private int calculateStringDumpRecordSize(Instance instance) {
+        return calculateInstanceDumpRecordSize(instance);
+    }
. . .
+    private int calculateThreadDumpRecordSize(Instance instance) {
+        return calculateInstanceDumpRecordSize(instance);
+    }
```
So, the methods `calculateStringDumpRecordSize()` and `calculateThreadDumpRecordSize()` are the same as the `calculateInstanceDumpRecordSize()`.

Then you have the code:
```
+        } else if (oop instanceof Instance) {
+            Instance instance = (Instance) oop;
+            Klass klass = instance.getKlass();
+            Symbol name = klass.getName();
+            if (name.equals(javaLangString)) {
+                return calculateStringDumpRecordSize(instance);
+            } else if (name.equals(javaLangClass)) {
+                return calculateClassInstanceDumpRecordSize(instance);
+            } else if (name.equals(javaLangThread)) {
+                return calculateThreadDumpRecordSize(instance);
+            } else {
+                klass = klass.getSuper();
+                while (klass != null) {
+                    name = klass.getName();
+                    if (name.equals(javaLangThread)) {
+                        return calculateThreadDumpRecordSize(instance);
+                    }
+                    klass = klass.getSuper();
+                }
+                return calculateInstanceDumpRecordSize(instance);
+            }
+        } else {
```
You can simplify the above with something like this:
```
+        } else if (oop instanceof Instance) {
+            Instance instance = (Instance) oop;
+            Klass klass = instance.getKlass();
+            Symbol name = klass.getName();
+            if (name.equals(javaLangString) ||
+                name.equals(javaLangClass) ||
+                name.equals(javaLangThread)) {
+                return calculateClassInstanceDumpRecordSize(instance);
+            } else {
+                return calculateInstanceDumpRecordSize(instance);
+            }
+        } else {
```
And my guess is that this simplified code can be simplified even further (Q: is it right?): 
```
+        } else if (oop instanceof Instance) {
+            Instance instance = (Instance) oop;
+                return calculateInstanceDumpRecordSize(instance);
+            }
+        } else {
```
And the methods `calculateStringDumpRecordSize()` and `calculateThreadDumpRecordSize()` can be removed as unneeded.

**Comment by linzang** [comment]
Hi @sspitsyn, 
Thanks for your review and comments, I will fix them asap.
BRs,
Lin

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java [hunk 3]
```diff
@@ -472,47 +468,192 @@ public synchronized void write(String fileName) throws IOException {
...
-            out.writeInt(0);
+    protected int calculateOopDumpRecordSize(Oop oop) throws IOException {
+        if (oop instanceof TypeArray) {
+            return calculatePrimitiveArrayDumpRecordSize((TypeArray)oop);
+        } else if (oop instanceof ObjArray) {
+            Klass klass = oop.getKlass();
+            ObjArrayKlass oak = (ObjArrayKlass) klass;
+            Klass bottomType = oak.getBottomKlass();
+            if (bottomType instanceof InstanceKlass ||
+                    bottomType instanceof TypeArrayKlass) {
```
Comment (sspitsyn) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java:495]: Formatting nit: I'd suggest to align `bottomType` with `bottomType`.

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java [hunk 4]
```diff
@@ -472,47 +468,192 @@ public synchronized void write(String fileName) throws IOException {
...
+            default:
+                throw new RuntimeException("should not reach here");
+        }
+        return size;
+    }
+
+    private int calculateClassInstanceDumpRecordSize(Instance instance) {
+        Klass reflectedKlass = java_lang_Class.asKlass(instance);
+        // dump instance record only for primitive type Class objects.
+        // all other Class objects are covered by writeClassDumpRecords.
```
Comment (sspitsyn) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java:611]: Formatting nit: If the comment is ended with dot then start it from capital a letter.

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java [hunk 5]
```diff
@@ -472,47 +468,192 @@ public synchronized void write(String fileName) throws IOException {
...
+                                                   headerSize,
+                                                   typeSize,
+                                                   typeName);
+        return headerSize + (int)typeSize * length;
     }
 
     @Override
-    protected void writeHeapRecordEpilogue() throws IOException {
+    protected void writeHeapRecordPrologue(int size) throws IOException {
+        if (size == 0 || (currentSegmentStart > 0)) {
```
Comment (sspitsyn) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java:642]: Formatting nit: There is no need in internal brackets `()` around `currentSegmentStart > 0`.

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java [hunk 6]
```diff
@@ -472,47 +468,192 @@ public synchronized void write(String fileName) throws IOException {
...
+            return;
+        }
+        // write heap data header
         if (useSegmentedHeapDump) {
-            hprofBufferedOut.exitSegmentMode();
+            out.writeByte((byte)HPROF_HEAP_DUMP_SEGMENT);
+            out.writeInt(0);
+            out.writeInt(size);
+        } else {
+            out.writeByte((byte) (HPROF_HEAP_DUMP));
```
Comment (sspitsyn) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java:651]: Formatting nit: No need in space after cast to `byte`.

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java [hunk 7]
```diff
@@ -472,47 +468,192 @@ public synchronized void write(String fileName) throws IOException {
...
+            out.writeInt(0);
+            // record position of the length slot
+            currentSegmentStart = fos.getChannel().position();
```
Comment (sspitsyn) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java:654]: The comment before this line is confusing: the `currentSegmentStart` is not the length.
Should this comment be moved after the line 653?

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java [hunk 8]
```diff
@@ -576,48 +717,26 @@ private int getArrayHeaderSize(boolean isObjectAarray) {
...
-                (currentRecordLength + headerSize + originalLengthInBytes) > MAX_U4_VALUE) {
-                fillInHeapRecordLength();
-                currentSegmentStart = 0;
-                writeHeapRecordPrologue();
-                currentRecordLength = 0;
-            }
-        }
         // Calculate the max bytes we can use.
-        long maxBytes = (MAX_U4_VALUE - (headerSize + currentRecordLength));
+        long maxBytes = (MAX_U4_VALUE - headerSize);
```
Comment (sspitsyn) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java:732]: Formatting nit: There is no need in brackets `()` here.

**Comment by sspitsyn** [comment]
Hi Lin,
Thank you for update!
I've added several comments with formatting nits.
Otherwise, it looks okay to me but I did not review everything deeply.
Thanks,
Serguei

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java (hunk 7 above)
Comment (linzang) [reply src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java:654]: Hi Serguei,
Here it gets the file position and record it as the start point of data written, this record would be used later for calculate the data written size which would be written in the file. 
I will update the comment.

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java (hunk 5 above)
Comment (linzang) [reply src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java:642]: Hi Serguei, 
The currentSegmentStart here could be non-zero when  useSegmentHeapDump is false. so we need to check it here. 
Thanks!

**Comment by linzang** [comment]
Dear Serguei @sspitsyn ，
Very appericate for your review and comments, I just updated the PR based on your comments. 
Thanks!
Lin

**Review by sspitsyn — APPROVED** [review]
Hi Lin,
Thank you for the update!
It looks good to me.
Thanks,
Serguei

**Comment by linzang** [comment]
Dear All,
May I ask your help to review? I believe this PR need one more approval. Thanks!

BRs,
Lin

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java [hunk 9]
```diff
@@ -472,47 +468,192 @@ public synchronized void write(String fileName) throws IOException {
...
+            // ik == null for instance fields
+            if (ik != null) {
+                // static field
+                size += calculateFieldDumpRecordSize(field, ik.getJavaMirror());
+            }
+        }
+        return size;
+    }
+
+    private int calculateFieldDumpRecordSize(Field field, Oop oop) throws IOException {
```
Comment (alexmenkov) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java:578]: This method is very similar to getSizeForFields
I think it makes sense to introduce method getSizeForField(Field) and use it here and in getSizeForFields
And I'd simplified the code to something like
```suggestion
    private int getSizeForField(Field field)  {
        char typeCode = (char)field.getSignature().getByteAt(0);
        switch (typeCode) {
            case JVM_SIGNATURE_BOOLEAN:
            case JVM_SIGNATURE_BYTE:
                return 1;
            case JVM_SIGNATURE_CHAR:
            case JVM_SIGNATURE_SHORT:
                return 2;
            case JVM_SIGNATURE_INT:
            case JVM_SIGNATURE_FLOAT:
                return 4;
            case JVM_SIGNATURE_CLASS:
            case JVM_SIGNATURE_ARRAY:
                return OBJ_ID_SIZE;
            case JVM_SIGNATURE_LONG:
            case JVM_SIGNATURE_DOUBLE:
                return 8;
            default:
                throw new RuntimeException("should not reach here");
        }
    }

```

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java [hunk 10]
```diff
@@ -1346,38 +1311,37 @@ private long getAddressValue(Address addr) {
         return res;
     }
 
+    // get size in bytes (in stream) required for given field.
+    private int getSizeForField(Field field) {
+        char typeCode = (char) field.getSignature().getByteAt(0);
```
Comment (sspitsyn) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java:1316]: Space after cast `(char)` is not not needed.

**Comment by sspitsyn** [comment]
Lin, thank you for the update. I like suggestions/simplifications from Alex.

**Review by alexmenkov — APPROVED** [review]

**Comment by linzang** [comment]
Thanks @sspitsyn and @alexmenkov for your review and approvel! I will push this PR after 1 day if there is no new comments. 

BRs,
Lin

**Pull request closed without merging**